File: dashboard/frontend/src/api/translators/entries.ts
import { actorId, entryId } from '../../app/domain-ids';
import type {
  Content,
  Entry,
  EntryContent,
  EntryPage,
  Question,
} from '../../entries/model';
import type { components } from '../generated/schema';

type Schemas = components['schemas'];

class EntryValidationFailure extends Error {
  readonly kind = 'validation';

  constructor(message: string) {
    super(message);
    this.name = 'EntryValidationFailure';
  }
}

function field(value: unknown, name: string): unknown {
  if (typeof value !== 'object' || value === null || !(name in value)) {
    throw new EntryValidationFailure(`entry field is missing: ${name}`);
  }
  return Reflect.get(value, name);
}

function text(value: unknown, name: string): string {
  const candidate = field(value, name);
  if (typeof candidate !== 'string') {
    throw new EntryValidationFailure(`entry field must be a string: ${name}`);
  }
  return candidate;
}

function optionalText(value: unknown, name: string): string | null {
  const candidate = field(value, name);
  if (candidate === null || typeof candidate === 'string') return candidate;
  throw new EntryValidationFailure(
    `entry field must be a string or null: ${name}`,
  );
}

function finiteNumber(value: unknown, name: string): number {
  const candidate = field(value, name);
  if (typeof candidate !== 'number' || !Number.isFinite(candidate)) {
    throw new EntryValidationFailure(
      `entry field must be a finite number: ${name}`,
    );
  }
  return candidate;
}

function optionalCount(value: unknown, name: string): number | null {
  const candidate = field(value, name);
  if (candidate === null) return null;
  if (typeof candidate === 'number' && Number.isFinite(candidate))
    return candidate;
  throw new EntryValidationFailure(
    `entry field must be a number or null: ${name}`,
  );
}

function flag(value: unknown, name: string): boolean {
  const candidate = field(value, name);
  if (typeof candidate !== 'boolean') {
    throw new EntryValidationFailure(`entry field must be a boolean: ${name}`);
  }
  return candidate;
}

function list(value: unknown, name: string): readonly unknown[] {
  const candidate = field(value, name);
  if (!Array.isArray(candidate)) {
    throw new EntryValidationFailure(`entry field must be an array: ${name}`);
  }
  return candidate;
}

function content(value: unknown): Content {
  const mediaType = text(value, 'media_type');
  if (mediaType !== 'text/plain' && mediaType !== 'text/markdown') {
    throw new EntryValidationFailure('entry has an unknown content media type');
  }
  return { text: text(value, 'text'), mediaType };
}

function optionalContent(value: unknown, name: string): Content | null {
  const candidate = field(value, name);
  return candidate === null ? null : content(candidate);
}

function oneOf<Value extends string>(
  value: unknown,
  name: string,
  choices: readonly Value[],
): Value {
  const candidate = text(value, name);
  const match = choices.find((choice) => choice === candidate);
  if (match === undefined) {
    throw new EntryValidationFailure(
      `entry field has an unknown value: ${name}`,
    );
  }
  return match;
}

function stringList(value: unknown, name: string): readonly string[] {
  return list(value, name).map((item) => {
    if (typeof item !== 'string') {
      throw new EntryValidationFailure(
        `entry list must contain strings: ${name}`,
      );
    }
    return item;
  });
}

function questions(value: unknown): readonly Question[] {
  return list(value, 'questions').map((question) => ({
    questionId: text(question, 'question_id'),
    title: optionalText(question, 'title'),
    question: text(question, 'question'),
    multiple: flag(question, 'multiple'),
    choices: list(question, 'choices').map((choice) => ({
      label: text(choice, 'label'),
      description: optionalText(choice, 'description'),
    })),
  }));
}

function messagePhase(
  value: unknown,
): 'prompt' | 'intermediate' | 'end_turn' | 'synthetic' | 'recap' | null {
  const candidate = optionalText(value, 'phase');
  switch (candidate) {
    case null:
    case 'prompt':
    case 'intermediate':
    case 'end_turn':
    case 'synthetic':
    case 'recap':
      return candidate;
    default:
      throw new EntryValidationFailure('entry has an unknown message phase');
  }
}

type EntryWireContent = {
  readonly type: EntryContent['type'];
  readonly body: unknown;
};

function entryType(value: unknown, name: string): EntryContent['type'] {
  const candidate = text(value, name);
  switch (candidate) {
    case 'turn_started':
    case 'turn_finished':
    case 'message':
    case 'reasoning':
    case 'shell_started':
    case 'shell_output':
    case 'shell_backgrounded':
    case 'shell_finished':
    case 'file':
    case 'search':
    case 'web':
    case 'worktree':
    case 'skill_started':
    case 'skill_finished':
    case 'question_asked':
    case 'question_answered':
    case 'plan_proposed':
    case 'plan_resolved':
    case 'compaction_started':
    case 'compaction_finished':
    case 'assignment_started':
    case 'assignment_finished':
    case 'model_change':
    case 'effort_change':
      return candidate;
    default:
      throw new EntryValidationFailure('entry has an unknown type');
  }
}

function entryContent(wire: EntryWireContent): EntryContent {
  const body: unknown = wire.body;
  switch (wire.type) {
    case 'turn_started':
      return { type: wire.type, body: {} };
    case 'turn_finished':
      return {
        type: wire.type,
        body: { state: oneOf(body, 'state', ['finished', 'aborted']) },
      };
    case 'message': {
      const recipient = optionalText(body, 'recipient_actor_id');
      return {
        type: wire.type,
        body: {
          messageId: text(body, 'message_id'),
          role: oneOf(body, 'role', [
            'user',
            'assistant',
            'system',
            'peer',
            'parent',
          ]),
          phase: messagePhase(body),
          content: content(field(body, 'content')),
          recipientActorId: recipient === null ? null : actorId(recipient),
          replyTo: optionalText(body, 'reply_to'),
        },
      };
    }
    case 'reasoning':
      return {
        type: wire.type,
        body: {
          reasoningId: text(body, 'reasoning_id'),
          content: content(field(body, 'content')),
        },
      };
    case 'shell_started':
      return {
        type: wire.type,
        body: {
          shellId: text(body, 'shell_id'),
          command: content(field(body, 'command')),
          execution: oneOf(body, 'execution', [
            'foreground',
            'background',
            'monitor',
          ]),
        },
      };
    case 'shell_output':
      return {
        type: wire.type,
        body: {
          shellId: text(body, 'shell_id'),
          stream: oneOf(body, 'stream', ['output', 'error', 'status']),
          mode: oneOf(body, 'mode', ['append', 'replace']),
          content: content(field(body, 'content')),
        },
      };
    case 'shell_backgrounded':
      return { type: wire.type, body: { shellId: text(body, 'shell_id') } };
    case 'shell_finished':
      return {
        type: wire.type,
        body: {
          shellId: text(body, 'shell_id'),
          state: oneOf(body, 'state', ['succeeded', 'failed', 'cancelled']),
          exitCode: optionalCount(body, 'exit_code'),
          result: optionalContent(body, 'result'),
        },
      };
    case 'file':
      return {
        type: wire.type,
        body: {
          path: text(body, 'path'),
          action: oneOf(body, 'action', [
            'read',
            'created',
            'updated',
            'deleted',
            'renamed',
          ]),
          state: oneOf(body, 'state', ['succeeded', 'failed']),
          previousPath: optionalText(body, 'previous_path'),
          linesAdded: optionalCount(body, 'lines_added'),
          linesRemoved: optionalCount(body, 'lines_removed'),
          content: optionalContent(body, 'content'),
        },
      };
    case 'search':
      return {
        type: wire.type,
        body: {
          tool: text(body, 'tool'),
          query: content(field(body, 'query')),
          state: oneOf(body, 'state', ['succeeded', 'failed']),
          result: optionalContent(body, 'result'),
        },
      };
    case 'web':
      return {
        type: wire.type,
        body: {
          url: optionalText(body, 'url'),
          state: oneOf(body, 'state', ['succeeded', 'failed']),
          result: optionalContent(body, 'result'),
        },
      };
    case 'worktree':
      return {
        type: wire.type,
        body: {
          action: oneOf(body, 'action', ['entered', 'exited']),
          state: oneOf(body, 'state', ['succeeded', 'failed']),
          arguments: optionalContent(body, 'arguments'),
        },
      };
    case 'skill_started':
      return {
        type: wire.type,
        body: {
          skillId: text(body, 'skill_id'),
          name: text(body, 'name'),
          arguments: optionalContent(body, 'arguments'),
        },
      };
    case 'skill_finished':
      return {
        type: wire.type,
        body: {
          skillId: text(body, 'skill_id'),
          state: oneOf(body, 'state', ['succeeded', 'failed', 'cancelled']),
          result: optionalContent(body, 'result'),
        },
      };
    case 'question_asked':
      return {
        type: wire.type,
        body: {
          attentionId: text(body, 'attention_id'),
          questions: questions(body),
        },
      };
    case 'question_answered':
      return {
        type: wire.type,
        body: {
          attentionId: text(body, 'attention_id'),
          answers: list(body, 'answers').map((answer) => ({
            questionId: text(answer, 'question_id'),
            labels: stringList(answer, 'labels'),
          })),
          feedback: optionalText(body, 'feedback'),
        },
      };
    case 'plan_proposed':
      return {
        type: wire.type,
        body: {
          attentionId: text(body, 'attention_id'),
          plan: content(field(body, 'plan')),
        },
      };
    case 'plan_resolved':
      return {
        type: wire.type,
        body: {
          attentionId: text(body, 'attention_id'),
          state: oneOf(body, 'state', [
            'approved',
            'changes_requested',
            'rejected',
          ]),
          feedback: optionalText(body, 'feedback'),
          edited: flag(body, 'edited'),
        },
      };
    case 'compaction_started':
      return {
        type: wire.type,
        body: { beforeTokens: optionalCount(body, 'before_tokens') },
      };
    case 'compaction_finished':
      return {
        type: wire.type,
        body: {
          beforeTokens: optionalCount(body, 'before_tokens'),
          afterTokens: optionalCount(body, 'after_tokens'),
        },
      };
    case 'assignment_started':
      return {
        type: wire.type,
        body: {
          assignmentId: text(body, 'assignment_id'),
          assignedActorName: optionalText(body, 'assigned_actor_name'),
          prompt: optionalContent(body, 'prompt'),
        },
      };
    case 'assignment_finished':
      return {
        type: wire.type,
        body: {
          assignmentId: text(body, 'assignment_id'),
          state: oneOf(body, 'state', ['succeeded', 'failed', 'cancelled']),
          result: optionalContent(body, 'result'),
        },
      };
    case 'model_change':
      return {
        type: wire.type,
        body: {
          current: text(body, 'current'),
          previous: optionalText(body, 'previous'),
          automatic: flag(body, 'automatic'),
        },
      };
    case 'effort_change':
      return {
        type: wire.type,
        body: {
          current: text(body, 'current'),
          previous: optionalText(body, 'previous'),
        },
      };
  }
}

export function translateEntry(wire: Schemas['EntryResponse']): Entry {
  const parent = wire.parent_actor_id;
  return {
    entryId: entryId(wire.entry_id),
    cursor: wire.cursor,
    actorId: actorId(wire.actor_id),
    parentActorId: parent === null ? null : actorId(parent),
    turnId: wire.turn_id,
    occurredAt: wire.occurred_at,
    summary: wire.summary,
    ...entryContent(wire),
  };
}

export function decodeEntry(value: unknown): Entry {
  const parent = optionalText(value, 'parent_actor_id');
  return {
    entryId: entryId(text(value, 'entry_id')),
    cursor: finiteNumber(value, 'cursor'),
    actorId: actorId(text(value, 'actor_id')),
    parentActorId: parent === null ? null : actorId(parent),
    turnId: optionalText(value, 'turn_id'),
    occurredAt: finiteNumber(value, 'occurred_at'),
    summary: optionalText(value, 'summary'),
    ...entryContent({
      type: entryType(value, 'type'),
      body: field(value, 'body'),
    }),
  };
}

export function translateEntryPage(
  wire: Schemas['EntryPageResponse'],
): EntryPage {
  return {
    items: wire.items.map(translateEntry),
    oldestCursor: wire.oldest_cursor,
    hasMore: wire.has_more,
  };
}

File: domain/entries.py
"""The session feed, as stored rows: one immutable entry per feed-worthy fact.

Built once, at push time, by `engine/sessiondata/entries.py`, and never touched
again — every entry is immutable, so materializing one is a pure append and
reading the feed is one indexed range scan. The canonical log stays behind them
as raw events and as the replay source.

An entry is NOT its canonical event. The event says what happened in the words
the translators produce; the entry says what a reader is shown, which drops
every field nobody displays (the reasons, the operation links, the harness's own
decision words) and folds the rest into one flat body per kind.

Content stays `Content` rather than a bare string: whether a message is markdown
or plain text is a fact about it, the harness told us, and a body that dropped
it would leave every client guessing by role.
"""

from __future__ import annotations

from collections.abc import Sequence
from dataclasses import dataclass
from enum import StrEnum

from domain.ids import (
    ActorId,
    AssignmentId,
    AttentionId,
    CanonicalEventId,
    MessageId,
    ReasoningId,
    SessionId,
    ShellId,
    SkillId,
    TurnId,
)
from domain.stored import STORED
from domain.values import (
    AttentionAnswer,
    AttentionPrompt,
    Content,
    ExecutionMode,
    FileAction,
    MessagePhase,
    MessageRole,
    OutputMode,
    PlanState,
    ProgressStream,
    WorktreeAction,
)


class EntryTypeName(StrEnum):
    """The kinds a feed has, as the name that travels with a stored row.
    Typed rather than `str` so the api layer's own copy of this vocabulary is
    CHECKED against it: two lists that must agree, and a type error the
    moment they do not."""

    TURN_STARTED = "turn_started"
    TURN_FINISHED = "turn_finished"
    MESSAGE = "message"
    REASONING = "reasoning"
    SHELL_STARTED = "shell_started"
    SHELL_OUTPUT = "shell_output"
    SHELL_BACKGROUNDED = "shell_backgrounded"
    SHELL_FINISHED = "shell_finished"
    FILE = "file"
    SEARCH = "search"
    WEB = "web"
    WORKTREE = "worktree"
    SKILL_STARTED = "skill_started"
    SKILL_FINISHED = "skill_finished"
    QUESTION_ASKED = "question_asked"
    QUESTION_ANSWERED = "question_answered"
    PLAN_PROPOSED = "plan_proposed"
    PLAN_RESOLVED = "plan_resolved"
    COMPACTION_STARTED = "compaction_started"
    COMPACTION_FINISHED = "compaction_finished"
    ASSIGNMENT_STARTED = "assignment_started"
    ASSIGNMENT_FINISHED = "assignment_finished"
    MODEL_CHANGE = "model_change"
    EFFORT_CHANGE = "effort_change"


class RunState(StrEnum):
    """How a thing that runs ended. One word for every kind, in place of the
    `outcome` + `reason` pair the canonical events carry: a feed shows the
    state, and nothing displayed the reason."""

    SUCCEEDED = "succeeded"
    FAILED = "failed"
    CANCELLED = "cancelled"


class TurnState(StrEnum):
    FINISHED = "finished"
    ABORTED = "aborted"


class FileState(StrEnum):
    SUCCEEDED = "succeeded"
    FAILED = "failed"


@dataclass(frozen=True)
class EntryBody:
    """Marker base for entry bodies. Same stored-shape rule as a payload: an
    unknown field in a decoded row is drift, not a new feature."""

    __pydantic_config__ = STORED


@dataclass(frozen=True)
class TurnStartedBody(EntryBody):
    """The grouping marker. Carries nothing: a turn's identity is on the
    stored event, and everything between two markers belongs to it."""


@dataclass(frozen=True)
class TurnFinishedBody(EntryBody):
    state: TurnState


@dataclass(frozen=True)
class MessageBody(EntryBody):
    message_id: MessageId
    role: MessageRole
    phase: MessagePhase | None
    content: Content
    recipient_actor_id: ActorId | None = None
    # The prompt this one replaced. A harness that re-parents around a DISCARDED
    # prompt still reports the dead one, and the only thing that tells the two
    # apart is that the survivor names the same parent — so a feed without this
    # shows a prompt nobody sent, for the rest of the session.
    reply_to: MessageId | None = None


@dataclass(frozen=True)
class ReasoningBody(EntryBody):
    reasoning_id: ReasoningId
    content: Content


@dataclass(frozen=True)
class ShellStartedBody(EntryBody):
    shell_id: ShellId
    command: Content
    execution: ExecutionMode


@dataclass(frozen=True)
class ShellOutputBody(EntryBody):
    """One chunk, exactly as it arrived. The client folds chunks per shell,
    which is a bounded fold over what is on screen — re-sending a growing
    output on every line was the churn this replaces."""

    shell_id: ShellId
    stream: ProgressStream
    mode: OutputMode
    content: Content


@dataclass(frozen=True)
class ShellBackgroundedBody(EntryBody):
    shell_id: ShellId


@dataclass(frozen=True)
class ShellFinishedBody(EntryBody):
    shell_id: ShellId
    state: RunState
    # Only ever filled by a harness that reports one, and not all of them do.
    exit_code: int | None = None
    # What the command PRINTED, when its harness reports the output at the END
    # rather than streaming it. Both channels exist because both kinds of harness
    # exist: one follows the output file and emits chunks as it grows, leaving
    # this empty; another reports the whole thing once, with the call's result,
    # and emits no chunks at all. A feed that read only the chunks showed every
    # command of the second kind with no output whatsoever.
    result: Content | None = None


@dataclass(frozen=True)
class FileBody(EntryBody):
    path: str
    action: FileAction
    state: FileState
    previous_path: str | None = None
    lines_added: int | None = None
    lines_removed: int | None = None
    # The diff when the file was changed, its text when it was created or read.
    content: Content | None = None


@dataclass(frozen=True)
class SearchBody(EntryBody):
    tool: str
    query: Content
    state: FileState
    result: Content | None = None


@dataclass(frozen=True)
class WebBody(EntryBody):
    url: str | None
    state: FileState
    result: Content | None = None


@dataclass(frozen=True)
class WorktreeBody(EntryBody):
    action: WorktreeAction
    state: FileState
    arguments: Content | None = None


@dataclass(frozen=True)
class SkillStartedBody(EntryBody):
    skill_id: SkillId
    name: str
    arguments: Content | None = None


@dataclass(frozen=True)
class SkillFinishedBody(EntryBody):
    skill_id: SkillId
    state: RunState
    result: Content | None = None


@dataclass(frozen=True)
class QuestionAskedBody(EntryBody):
    """Pending until its answered twin arrives — which is how every client
    derives "this session is waiting on me", with no stored flag to go stale."""

    attention_id: AttentionId
    questions: tuple[AttentionPrompt, ...]


@dataclass(frozen=True)
class QuestionAnsweredBody(EntryBody):
    attention_id: AttentionId
    answers: tuple[AttentionAnswer, ...] = ()
    feedback: str | None = None


@dataclass(frozen=True)
class PlanProposedBody(EntryBody):
    attention_id: AttentionId
    plan: Content


@dataclass(frozen=True)
class PlanResolvedBody(EntryBody):
    attention_id: AttentionId
    state: PlanState
    feedback: str | None = None
    edited: bool = False


@dataclass(frozen=True)
class CompactionStartedBody(EntryBody):
    before_tokens: int | None = None


@dataclass(frozen=True)
class CompactionFinishedBody(EntryBody):
    before_tokens: int | None = None
    after_tokens: int | None = None
    context: Content | None = None


@dataclass(frozen=True)
class AssignmentStartedBody(EntryBody):
    assignment_id: AssignmentId
    assigned_actor_name: str | None = None
    prompt: Content | None = None


@dataclass(frozen=True)
class AssignmentFinishedBody(EntryBody):
    assignment_id: AssignmentId
    state: RunState = RunState.SUCCEEDED
    result: Content | None = None


@dataclass(frozen=True)
class ModelChangeBody(EntryBody):
    """`automatic` marks a harness fallback — the one `reason` that survived,
    because a model the harness chose for you is worth a warning."""

    current: str
    previous: str | None = None
    automatic: bool = False


@dataclass(frozen=True)
class EffortChangeBody(EntryBody):
    current: str
    previous: str | None = None


ENTRY_TYPES: dict[type[EntryBody], EntryTypeName] = {
    TurnStartedBody: EntryTypeName.TURN_STARTED,
    TurnFinishedBody: EntryTypeName.TURN_FINISHED,
    MessageBody: EntryTypeName.MESSAGE,
    ReasoningBody: EntryTypeName.REASONING,
    ShellStartedBody: EntryTypeName.SHELL_STARTED,
    ShellOutputBody: EntryTypeName.SHELL_OUTPUT,
    ShellBackgroundedBody: EntryTypeName.SHELL_BACKGROUNDED,
    ShellFinishedBody: EntryTypeName.SHELL_FINISHED,
    FileBody: EntryTypeName.FILE,
    SearchBody: EntryTypeName.SEARCH,
    WebBody: EntryTypeName.WEB,
    WorktreeBody: EntryTypeName.WORKTREE,
    SkillStartedBody: EntryTypeName.SKILL_STARTED,
    SkillFinishedBody: EntryTypeName.SKILL_FINISHED,
    QuestionAskedBody: EntryTypeName.QUESTION_ASKED,
    QuestionAnsweredBody: EntryTypeName.QUESTION_ANSWERED,
    PlanProposedBody: EntryTypeName.PLAN_PROPOSED,
    PlanResolvedBody: EntryTypeName.PLAN_RESOLVED,
    CompactionStartedBody: EntryTypeName.COMPACTION_STARTED,
    CompactionFinishedBody: EntryTypeName.COMPACTION_FINISHED,
    AssignmentStartedBody: EntryTypeName.ASSIGNMENT_STARTED,
    AssignmentFinishedBody: EntryTypeName.ASSIGNMENT_FINISHED,
    ModelChangeBody: EntryTypeName.MODEL_CHANGE,
    EffortChangeBody: EntryTypeName.EFFORT_CHANGE,
}

BODY_TYPES: dict[EntryTypeName, type[EntryBody]] = {
    entry_type: body for body, entry_type in ENTRY_TYPES.items()
}


ATTENTION_ENTRY_TYPES: tuple[EntryTypeName, ...] = (
    EntryTypeName.QUESTION_ASKED,
    EntryTypeName.QUESTION_ANSWERED,
    EntryTypeName.PLAN_PROPOSED,
    EntryTypeName.PLAN_RESOLVED,
)


@dataclass(frozen=True)
class SessionEntry:
    """One row of `session_entries`.

    `entry_id` is the source canonical event's id, which is what makes writing
    one idempotent: a replay inserts the same row and the UNIQUE constraint
    keeps it at one. `cursor` is the read model's own monotonic stamp — the SSE
    event id, and the paging key.
    """

    entry_id: CanonicalEventId
    session_id: SessionId
    actor_id: ActorId
    parent_actor_id: ActorId | None
    turn_id: TurnId | None
    occurred_at: float
    summary: str | None
    body: EntryBody
    cursor: int = 0

    @property
    def entry_type(self) -> EntryTypeName:
        return ENTRY_TYPES[type(self.body)]


def pending_attention(entries: Sequence[SessionEntry]) -> tuple[SessionEntry, ...]:
    """The questions and plans still waiting on a person, oldest first.

    DERIVED, from the attention entries in order: an asked or proposed entry
    whose resolution has not arrived. Nothing stores "still pending" — a stored
    flag would be a second answer to the same question, and it could disagree
    with the feed the person is looking at.

    The same rule both clients apply to draw the attention badge, so it lives
    here rather than three times.
    """
    open_attentions: dict[AttentionId, SessionEntry] = {}
    for entry in entries:
        body = entry.body
        if isinstance(body, (QuestionAskedBody, PlanProposedBody)):
            open_attentions[body.attention_id] = entry
        elif isinstance(body, (QuestionAnsweredBody, PlanResolvedBody)):
            open_attentions.pop(body.attention_id, None)
    return tuple(open_attentions.values())
